src/vos/vos_tree.c

```diff
@@ -963,18 +995,21 @@ key_tree_punch(struct vos_object *obj, daos_handle_t toh, daos_iov_t *key_iov,
 		daos_iov_t	         tmp;
 
 		krec2 = rbund->rb_krec;
-		krec2->kr_btr = krec->kr_btr;
-		umem_tx_add_ptr(umm, &krec->kr_btr, sizeof(krec->kr_btr));
-		memset(&krec->kr_btr, 0, sizeof(krec->kr_btr));
-
-		if (krec->kr_bmap & KREC_BF_EVT) {
-			krec2->kr_evt[0] = krec->kr_evt[0];
-			umem_tx_add_ptr(umm, &krec->kr_evt[0],
-					sizeof(krec->kr_evt[0]));
-			memset(&krec->kr_evt[0], 0, sizeof(krec->kr_evt[0]));
+		if (krec->kr_bmap & KREC_BF_BTR) {
+			krec2->kr_btr = krec->kr_btr;
+			umem_tx_add_ptr(umm, &krec->kr_btr,
+					sizeof(krec->kr_btr));
+			memset(&krec->kr_btr, 0, sizeof(krec->kr_btr));
+		} else {
+			D_ASSERT(krec->kr_bmap & KREC_BF_EVT);
+			krec2->kr_evt = krec->kr_evt;
+			umem_tx_add_ptr(umm, &krec->kr_evt,
+					sizeof(krec->kr_evt));
+			memset(&krec->kr_evt, 0, sizeof(krec->kr_evt));
 		}
 		/* Capture the earliest modification time of the removed key */
 		krec2->kr_earliest = krec->kr_earliest;
+		krec2->kr_bmap = krec->kr_bmap | KREC_BF_PUNCHED;
```

Comment:
Why need the "KREC_BF_PUNCHED" flag? it should have been set when ktr_rec_alloc().

Reply:
for krec2, yes.  But the removed key will not have the punched flag set.   We could potentially do bmap |= krec->kr_bmap and have same effect.   But I don't see much value in that.

Reply:
The key with non-DAOS_EPOCH_MAX epoch will be assigned KREC_BF_PUNCHED when it is allocated, it is the current "krec2" that will be kept in the tree after the punch. The key with DAOS_EPOCH_MAX will be removed from the tree after the punch, so it does not need the KREC_BF_PUNCHED flag.

Reply:
Right.   I'm changing krec2's bmap here which is the one that, as you say, already has it set.

I can either do
krec2->kr_bmap |= krec->kr_bmap

or

krec2->kr_bmap = krec->kr_bmap | KREC_BF_PUNCHED

Both statemetns are equivalent.   In one case, I'm preserving the flag that is already set and in the other, I'm just adding it back.

The one I have may be slightly more efficient since it's only doing a write to krec2->kr_bmap whereas, the first one above would require reading it first.   In this case, we know it's punched so I opted for the one where I set the flag again.   The main thing here is that the bmap needs to inherit the flags set in krec->kr_bmap, particularly relating to the type of tree